src/lib/tabbed.c

```diff
@@ -0,0 +1,433 @@
+#include "internal.h"
+
+void nctabbed_redraw(nctabbed* nt){
+  nctab* t;
+  int drawn_cols = 0;
+  int rows, cols;
+  if(nt->tabcount == 0){
+    // no tabs = nothing to draw
+    ncplane_erase(nt->hp);
+    return;
+  }
+  // update sizes for planes
+  ncplane_dim_yx(nt->ncp, &rows, &cols);
+  if(nt->opts.flags & NCTABBED_OPTION_BOTTOM){
+    ncplane_resize_simple(nt->hp, -1, cols);
+    ncplane_resize_simple(nt->p, rows - 1, cols);
+    ncplane_move_yx(nt->hp, rows - 2, 0);
+  }else{
+    ncplane_resize_simple(nt->hp, -1, cols);
+    ncplane_resize_simple(nt->p, rows - 1, cols);
+  }
+  // the callback draws the tab contents
+  if(nt->selected->cb){
+    nt->selected->cb(nt->selected, nt->p, nt->selected->curry);
+  }
+  // now we draw the headers
+  t = nt->leftmost;
+  ncplane_erase(nt->hp);
+  ncplane_set_channels(nt->hp, nt->opts.hdrchan);
+  do{
+    if(t == nt->selected){
+      ncplane_set_channels(nt->hp, nt->opts.selchan);
+      drawn_cols += ncplane_putstr(nt->hp, t->name);
+      ncplane_set_channels(nt->hp, nt->opts.hdrchan);
+    }else{
+      drawn_cols += ncplane_putstr(nt->hp, t->name);
+    }
+    // avoid drawing the separator after the last tab, or when we
+    // ran out of space, or when it's not set
+    if((t->next != nt->leftmost || drawn_cols >= cols) && nt->opts.separator){
+      ncplane_set_channels(nt->hp, nt->opts.sepchan);
+      drawn_cols += ncplane_putstr(nt->hp, nt->opts.separator);
+      ncplane_set_channels(nt->hp, nt->opts.hdrchan);
+    }
+    t = t->next;
+  }while(t != nt->leftmost && drawn_cols < cols);
+}
+
+void nctabbed_ensure_selected_header_visible(nctabbed* nt){
+  nctab* t = nt->leftmost;
+  int cols = ncplane_dim_x(nt->hp);
+  int takencols = 0;
+  if(!t){
+    return;
+  }
+//fprintf(stderr, "ensuring selected header visible\n");
+  do{
+    if(t == nt->selected){
+      break;
+    }
+    takencols += t->namecols + nt->sepcols;
+    if(takencols >= cols){
+//fprintf(stderr, "not enough space, rotating\n");
+      takencols -= nt->leftmost->namecols + nt->sepcols;
+      nctabbed_rotate(nt, -1);
+    }
+    t = t->next;
+//fprintf(stderr, "iteration over: takencols = %d, cols = %d\n", takencols, cols);
+  }while(t != nt->leftmost);
+//fprintf(stderr, "ensuring done\n");
+}
+
+static bool
+nctabbed_validate_opts(ncplane* n, const nctabbed_options* opts){
+  if(!n){
+    return false;
+  }
+  if(opts->flags > NCTABBED_OPTION_BOTTOM){
```

Comment:
probably want a `logerror()` on each of these. if i'm missing such in whatever you cribbed this from, let me know and i'll go add them there.

Reply:
actually, invalid flags ought only be a warning, so don't `return -1` here (and move it to the top, so the warning is always seen even if other checks fail). again, if i have counterbehavior that you cribbed from, let me know and i'll go fix it there.

Reply:
I based this off the ncreel code, which doesn't call `logerror()`. It does warn on invalid flags though. I guess I'll do what newer widgets do. ncreel's options validation code should be changed then?

Reply:
yep i'm updating it now. for some historical reference, the reel code existed before notcurses ever did, and indeed has been around as a coherent whole for almost a decade. indeed, most of the documentation comes right from here: https://nick-black.com/dankwiki/index.php?title=Outcurses#Panelreels so it wasn't written ground-up as notcurses code.

Reply:
actually hrmm

```c
static bool
validate_ncreel_opts(ncplane* n, const ncreel_options* ropts){
  if(n == NULL){
    return false;
  }
  if(ropts->flags >= (NCREEL_OPTION_CIRCULAR << 1u)){
    logwarn(ncplane_notcurses(n), "Provided unsupported flags 0x%016jx\n", (uintmax_t)ropts->
  }
  if(ropts->flags & NCREEL_OPTION_CIRCULAR){
    if(!(ropts->flags & NCREEL_OPTION_INFINITESCROLL)){
      return false; // can't set circular without infinitescroll
    }
  }
```

so it shows how to do the options check correctly but yeah is missing `logerror()` elsewhere.

Reply:
ea09089b8

Reply:
but should i check for the ncplane being NULL?

Reply:
ahhhh the reason i wasn't checking for `NULL` there is because these are marked `__attribute__ ((nonull (1)))`. those non-`NULL` checks might get elided out by the compiler, and gcc even used to warn when such checks were used with `nonnull`. it doesn't appear to do so anymore, though, so i've added one to `tree` because why not. if we hit a compiler that dosn't understand `nonnull`, it'll be good to have. 57f8dc439

Reply:
i don't print errors there because it's more of a blunder than anything else...i don't see how such messages could be useful to the end user. that's kind of a thin argument, though. i think i just don't because i expect such checks to be elided, honestly. not a good argument at all. go with your heart about whether there ought be a `logerror()` on the `NULL` check.

btw did you add any unit tests (`src/tests`)? no worries if not.

Reply:
I didn't add any tests, no. I wanted to, but I don't know what the tests exactly should test. Should they validate what the widget is drawing, or just what different functions take/return and how they behave?

Reply:
whatever you can =] the latter gets to be kind of a pain (you'll want to look at `ncplane_at_yx()` and `notcurses_at_yx()`). unit tests are the kind of thing you can spend unlimited time on, so just do whatever you feel appropriate. you'll be able to run them with `notcurses-tester`, which is run as part of `make test` (which invokes `ctest`).

Reply:
will do.